### PR TITLE
Hashing: merkle_verify_proof

### PR DESCRIPTION
**Pull request by mationorato**
Add merkle proof verifier to hashing library

- [x] code merkle_verify_proof() function
- [x] add documentation
- [x] add unit tests

### CODE REVIEW

File: std/src/main.leo [hunk 1]
```diff
@@ -46,6 +46,32 @@ function test_merkle_root_is_different () {
...
+function test_proofs_validation(){
+    let leave: field = 1;
+    let proof: [field;3] = [2, 5639452814959479716333797092602429568041408410490188951967730026708667799362,6811329200201340627416984918056766552180813916433450573759408907577084397184];
+    let path: [bool;3] = [false,false,false];
+    let root: field = 833453432365499780361700154225551479355839019916612454159711825224580327701;
+    console.assert(verify_proof(leave, proof,path , root));
+    let bad_proof: [field;3] = [3, 5639452814959479716333797092602429568041408410490188951967730026708667799362,6811329200201340627416984918056766552180813916433450573759408907577084397184];
+    console.assert(!verify_proof(leave,bad_proof,path, root));
+    let bad_path: [bool;3] = [true,true,true];
+    console.assert(!verify_proof(leave,proof,bad_path, root));
```
Comment (Loco1982) [inline std/src/main.leo:61]:
```suggestion
    console.assert(!verify_proof(leave, proof, bad_path, root));
```

File: README.md [hunk 2]
```diff
@@ -85,6 +85,14 @@ merkle_root(fields_array: [field; 8]) -> field
 Computes a merkle root from an fixed 8 element array of fields.
 This implementation uses the poseidon algorithm for hashing.
 
+```
+merkle_verify_proof(leave:field, proof: [field;8], path_index: [bool;n], root: field) -> bool
```
Comment (pablodeymo) [inline README.md:89]: I think there is a typo in the name of the argument `leave`

**Review by pablodeymo — CHANGES_REQUESTED** [review]

File: std/src/hashing.leo [hunk 3]
```diff
@@ -286,3 +286,17 @@ function merkle_root(array: [field;8]) -> field {
...
     return array_of_leaves[0];
 }
+
+function verify_proof(leave:field, proof: [field;3], path_index: [bool;3], root: field) -> bool{
+    const levels: u32 = 3;
+    
+    let element = leave;
+    for i in 0..levels{
+        if path_index[i] {
+            element = poseidon(proof[i],element);
```
Comment (Loco1982) [inline std/src/hashing.leo:296]:
```suggestion
            element = poseidon(proof[i], element);
```

File: std/src/main.leo [hunk 4]
```diff
@@ -46,6 +46,32 @@ function test_merkle_root_is_different () {
     console.assert(markle_root_1!=markle_root_2);
 }
 
+// verify_proof() tests
+
+@test()
+function test_proofs_validation(){
+    let leave: field = 1;
+    let proof: [field;3] = [2, 5639452814959479716333797092602429568041408410490188951967730026708667799362,6811329200201340627416984918056766552180813916433450573759408907577084397184];
+    let path: [bool;3] = [false,false,false];
```
Comment (Loco1982) [inline std/src/main.leo:55]:
```suggestion
    let path: [bool;3] = [false, false, false];
```

File: std/src/main.leo [hunk 5]
```diff
@@ -46,6 +46,32 @@ function test_merkle_root_is_different () {
...
+    let root: field = 833453432365499780361700154225551479355839019916612454159711825224580327701;
+    console.assert(verify_proof(leave, proof,path , root));
+    let bad_proof: [field;3] = [3, 5639452814959479716333797092602429568041408410490188951967730026708667799362,6811329200201340627416984918056766552180813916433450573759408907577084397184];
+    console.assert(!verify_proof(leave,bad_proof,path, root));
+    let bad_path: [bool;3] = [true,true,true];
```
Comment (Loco1982) [inline std/src/main.leo:60]:
```suggestion
    let bad_path: [bool;3] = [true, true, true];
```

File: std/src/main.leo [hunk 6]
```diff
@@ -46,6 +46,32 @@ function test_merkle_root_is_different () {
...
+    console.assert(!verify_proof(leave,proof,bad_path, root));
+    let bad_root: field = 32;
+    console.assert(!verify_proof(leave,proof,path, bad_root));
```
Comment (Loco1982) [inline std/src/main.leo:63]:
```suggestion
    console.assert(!verify_proof(leave, proof, path, bad_root));
```

File: std/src/main.leo [hunk 7]
```diff
@@ -46,6 +46,32 @@ function test_merkle_root_is_different () {
...
+
+    let leave2: field = 3;
+    let proof2: [field;3] = [2, 6741600571531499538076626459614881881572634471432793830237600055083617801110,7040568090966913659889165548176415317933794382062127083870106037652670242975];
+    let path2: [bool;3] = [true,false,true];
```
Comment (Loco1982) [inline std/src/main.leo:67]:
```suggestion
    let path2: [bool;3] = [true, false, true];
```

**Review by Loco1982 — CHANGES_REQUESTED** [review]
Minor spelling changes

File: README.md (hunk 2 above)
Comment (MauroToscano) [reply README.md:89]: You are right! I'm fixing it

**Pull request merged**